### PR TITLE
Sketch: function calls faster than fastcall

### PR DESCRIPTION
**Pull request by stepancheg**
TL;DR: instead of look parameter by name during each function call,
cache argument to parameter mapping by call signature.

`class UltraFastCallSig` is a pair of (`numPositional`, `named`)
argument which can be used in callable implementation to cache
parameter resolving.

For invocation:

```
foo(1, 2, 4, c=3, d=4)
```

`UltraFastCallSig` is `(4, ["c", "d"])` regardless of what `foo` is.

`UltraFastCallSig` is only used in calls without varargs. In could
be extended to calls with varargs, but it would be larger change
than this PR.

`StarlarkFunction` implementation now contains a `ConcurrentHashMap`
from `UltraFastCallSig` to a mapping.

The mapping is an integer array, describing, how to populate arguments
after positional arguments (from given indexed named argument, or
from default value).

When mapping exists, it means that this particular call signature
is compatible with this function signature, and parameter population
no longer performs map lookups.

`UltraFastCallSig` can be also used to efficiently generate
`BuiltinFunction` invocations, but that's a story for another PR.

**This PR is a sketch**. I'd like to get some feedback out it before
spending more time on it.

Results for this silly test case:

```
def bar(x, y, z, w, a, b, c):
    pass

def test():
    for i in range(10):
        print(i)
        for j in range(1000):
            for j in range(100):
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
                bar(x=1, y="a", z=1, w=2, a="x", b="y", c="c")
test()
```

```
Before: 5.6s
After:  4.7s
```

The optimization is not applied for positional-only calls. For this benchmark:

```
def foo():
    pass

def test():
    for i in range(10):
        print(i)
        for j in range(1000):
            for k in range(200):
                foo()
                foo()
                foo()
                foo()
                foo()
                foo()
                foo()
                foo()
                foo()
                foo()

test()
```

The slowdown is within 1%.

Finally, this optimization has a benefit even single-name arg calls:

```
def foo(a):
    pass

def test():
    for i in range(10):
        print(i)
        for j in range(1000):
            for k in range(200):
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)
                foo(a=1)

test()
```

```
A: N=34, r=3.280+-0.077
B: N=34, r=3.222+-0.104
B/A: 0.982
```

### CODE REVIEW

**Comment by stepancheg** [comment]
@alandonovan?
